Serialize scraper

octopusinvitro opened the pull request:
This PR rewrites the actual behaviour of the scraper to use the scraped page gem. I moved the code from the scraper to the relevant classes.

Since the data of a politician was half in the members page and half in the dedicated member page, the scraping of the entry is represented by a class as well as the scraping of the dedicated member's page, its hash merged with the dedicated page hash.

# Checklists

* [x] scraper is on Morph.io under the "everypolitician-scrapers" group <https://morph.io/everypolitician-scrapers/zambia-na-scraper>
* [x] scraper's GitHub "Website" link points at morph.io page <https://github.com/everypolitician-scrapers/zambia-na-scraper>
* [x] scraper is set to auto-run <https://morph.io/everypolitician-scrapers/zambia-na-scraper/settings>
* [x] scraper is configured for archiving _(unless source doesn't require that)_
* [ ] ~legislature has a scraper webhook set _(usually only set on Wikidata person-data scrapers)_~ Not needed


## Archiving

* [x] we are using at least version 0.5 of scraped-page-archive-gem
* [x] scraper uses scraped-page-archive gem directly or via a suitable strategy
* [x] MORPH_SCRAPER_CACHE_GITHUB_REPO_URL is configured
* [x] pages are being archived in new branch of correct scraper repo <https://github.com/everypolitician-scrapers/zambia-na-scraper/tree/scraped-pages-archive/www.parliament.gov.zm>

## Gemfile

* [x] all links are secure
* [x] links to Github use `github:` protocol, not simply `git:`
* [x] file passes Rubocop with our normal setup (in particular, single quotes rather than double)

octopusinvitro committed:
Add Gemfile as the scraper didn't had one

The Gemfile was copied from the austria-parliament repository
https://github.com/everypolitician-scrapers/austria-parlament

It also uses the new scraped page archive

octopusinvitro committed:
Add Gemfile.lock after adding Gemfile

octopusinvitro committed:
Copy rubocop config from austria-parlament repo

Blanked the rubocop TODO file because we're starting a new project and it's
best to start clean

octopusinvitro committed:
Use an instance of a ScraperPage child class in the scraper

octopusinvitro committed:
Add a method to get the member's url

For now passing entry as an argument, this will be cleaned up later.

octopusinvitro committed:
Add methods to get the party_name and party_id

For now we are passing the entry, this will be cleaned up in future
commits.

octopusinvitro committed:
Add a method to get the member's name

Passing the entry for now, this will be cleaned in future commits.

octopusinvitro committed:
Add a method to get the member's photo

For now passing entry as an argument, this will be cleaned up later.

octopusinvitro committed:
Add a method to get the member's constituency

For now passing entry as an argument, this will be cleaned up later.

octopusinvitro committed:
Add a member page class with an id field and use it in the scraper

octopusinvitro committed:
Add email field to member page class

octopusinvitro committed:
Add birth date to member page class

octopusinvitro committed:
Refactor the scraper code to eliminate variables and align fields

octopusinvitro committed:
Move the name scraping to the member page

Each entry box from the members page is passed to the member page
as a secondary source for scraping, since the information for a member
is part in the members page, part in the member page.

octopusinvitro committed:
Move the photo scraping to the member page as a field

octopusinvitro committed:
Move the constituency scraping to the member page as a field

octopusinvitro committed:
Move the party information scraping to the member page as a field

octopusinvitro committed:
Move the source and term scraping to the member page as fields

octopusinvitro committed:
Remove unused code

tmtmtmtm commented on Gemfile:
```diff
@@ -0,0 +1,17 @@
+gem 'require_all'
+gem 'rubocop'
+gem 'scraped_page_archive', github: 'everypolitician/scraped_page_archive', branch: 'version-0.5.0'
+gem 'scraped_page', github: 'everypolitician/scraped_page', branch: 'prototype'
```
Do these need to be off branches?

tmtmtmtm commented on lib/member_page.rb:
```diff
@@ -0,0 +1,78 @@
+require 'scraped_page'
+
+class MemberPage < ScrapedPage
+  def initialize(url:, strategy: Strategy::LiveRequest.new, entry:)
```
Having to override the `initialize` method is a warning sign that this probably isn't the best way to go about this. 

It's also not clear what `entry` is in this context (i.e. that might be a plausible name for this in the caller, but it doesn't seem to convey much here)

tmtmtmtm commented on lib/members_page.rb:
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+require 'scraped_page'
+
+class MembersPage < ScrapedPage
```
Having this be a subclass of ScrapedPage, but not have any fields defined is a bit of a warning sign. Having fields defined in a different class seems even more so.

I suspect things that come from this page should be defined as fields in this class, and then something else should combine them later.

tmtmtmtm commented on scraper.rb:
```diff
@@ -1,68 +1,38 @@
+base = 'http://www.parliament.gov.zm'
 
-# We should really extract these from the 'Next' links…
+# We should really extract these from the 'Next' links...
```
The new layout looks like it has more pages than this now.

octopusinvitro committed:
Add a member entry class to represent each member on a list

This class will contain the fields that are present in an entry of a
members list in the members page, and will be merged with the
fields that are present in the dedicated member page.

octopusinvitro committed:
Move fields from member page class to the member entry class

These fields live really in the members page class. Since the member
entry class represents each entry in the list of members in the members
page, they were moved to the member entry class.

octopusinvitro committed:
Rewrite the mp_entries method as a field

octopusinvitro commented:
@tmtmtmtm Feedback addressed.
- The MembersPage class inherits from ScrapedPage so that it can use the archiver when opening a url.
- I created a new class for a member entry, that doesn't need to inherit from ScrapedPage but includes the FieldSerializer module.
- Since this PR is just to rewrite the scraping behaviour I would leave updating the urls to scrape for a different PR.

tmtmtmtm commented:
@octopusinvitro as discussed numerous times before, please stop rewriting history mid-review. If your new commits had simply been added after the prior review, then it would be possible to look at them in isolation. But when you rewrite all the earlier commits like this, then a reviewer needs to go back and examine them **all** again, adding a significant amount of time and effort to the process.

I also don't see a lot of point in re-reviewing at a code level yet, when you're still not even fetching all the members (e.g. http://www.parliament.gov.zm/node/411), and I don't understand your point about why you aren't doing that.

But without looking further at the code, and only looking at the output, I note:
- the `source` field for each person is set to a list page, rather than a personal page
- the `term` is set to '2011' when this is now the 2016 term,
- the `image` would be better as the large ones from the person page rather than the small ones from the list page
- this is only picking up a single email address even when more than one is listed (e.g. on http://www.parliament.gov.zm/node/4066)
- this isn't picking up any of the other contact information that's available (Phone, Mobile, Fax, Address).

octopusinvitro commented:
@tmtmtmtm Sorry for rebasing, I thought it would help the flow of commits to be more readable, I won't do it again.

Regarding your comment on `source`, `term`, `image`, `email` and contact details, that's what was in the scraper already, [as well as in the data](https://cdn.rawgit.com/everypolitician/everypolitician-data/1998d0cdc4bbae94914d3a34e78854c67a026597/data/Zambia/Assembly/ep-popolo-v1.0.json) (the source field for each person is set to a list page, rather than a personal page, they have a single email address, etc.).

As I said in the description, this PR was just for moving the existing code into classes, without making any extra changes (**_make the change easy_**) to then open a second PR where I would update the urls and the information for the current term (**_make the easy change_**). I believe that, had I done those changes in this PR, I would have been told to split the PR in two.

I am happy to add those changes to this PR tomorrow, unless you prefer that I leave this PR in pause for this week and instead work on adding the archiver to the scrapers with forthcoming elections.

tmtmtmtm commented:
@octopusinvitro that approach, whilst often useful, only works when the original code is actually working, and so continues to be working at the end of each preparatory stage.

Here you're taking a scraper that was written for the previous term, and rewriting it for a new term, when the underlying website it's scraping has changed, not only in page layout, but even in where the pages live. As such, simply rewriting the old scraper to do the same thing in a different way is not only of no value, but is actively dangerous, as putting something like this live would _look_ like it's working, but would actually be silently ignoring members, etc. To a certain extent that could potentially be mitigated by ensuring that there's a ticket reminding us to go back and pick those up before putting this 'live' (as this has moved location, and would need the `instructions.json` updated to point at it), but you haven't linked to any such ticket, and, anyway, that's not really a good way to break the work down.

This is essentially an entirely new codebase, even though it's disguised by being over the core of an existing one. So, the better way to break it down would be the same way as breaking down other new scrapers: first write the version that archives each of the pages we want, but doesn't do any other data extraction, thus getting the basic class structure etc in place. Then, in a second PR, extract all the fields and save to morph. Or, if _that_ change becomes too big, make the second PR be for adding the public API to all the data in the ScrapedPage classes (which could in turn be split into multiple PRs if needed), and have a final PR to then get that data in scraper.rb and write it to the database.

However, yes, adding archiving to sites that might disappear quite soon, either because they have an upcoming election, or have recently had one but not updated the site yet, is of higher priority this week.

octopusinvitro committed:
Add a method to get next page to scrape and use it in the scraper

octopusinvitro committed:
Update term with the correct value

octopusinvitro committed:
Add photo from the member's page instead of from the members' page

octopusinvitro committed:
Include all emails listed in member's page

octopusinvitro committed:
Pick contact information that's available

Phone, Mobile, Fax, Address
